### PR TITLE
Wrapped to wrapper block maping + rpc

### PR DESCRIPTION
**Pull request by tdelabro**
# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

No RPC server for starknet pallet

# What is the new behavior?

- a rpc server for starknet pallet
- storing block during the on_finalize execution also add a log to the digest
- the digest is read by a new syncing process that stores a mapping from starknet block hashes to substrate block hashes in a new database. This allows a O(1) complexity when calling the RPC server using the starknet hash.
- an override functionality has also been introduced. It allow to query the storage directly, without using the runtime API, while retaining the ability to fallback on it if needed.

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

# Other information

The whole architecture and abstractions follow really closely the frontier implementation, for good reasons.

### CODE REVIEW

**Review by abdelhamidbakhta — CHANGES_REQUESTED** [review]
Left some comments. Most of them are about missing Rust doc. Basically please add Rust doc in all added struct, traits, functions with the conventions we use in the codebase. I feel a bit bad with the facts that there are no tests but I understand they are not trivial to implement and we have a short time to meet the deadline for our demo goal. I am enclined to accept the PR with the missing tests but please add an issue about adding them later. Also I would love if you could do a live walkthrough on this PR to explain the architecture. I think it would be super useful content for new contributors that are not yet familiar with substrate.

File: crates/client/db/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    pub const CURRENT_SYNCING_TIPS: &[u8] = b"CURRENT_SYNCING_TIPS";
+}
+
+pub struct Backend<Block: BlockT> {
+    meta: Arc<MetaDb<Block>>,
+    mapping: Arc<MappingDb<Block>>,
+}
+
+/// Returns the frontier database directory.
+pub fn frontier_database_dir(db_config_dir: &Path, db_path: &str) -> PathBuf {
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/db/src/lib.rs]: ✅

File: crates/client/db/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    db_config_dir.join("starknet").join(db_path)
+}
+
+impl<Block: BlockT> Backend<Block> {
+    pub fn open(database: &DatabaseSource, db_config_dir: &Path) -> Result<Self, String> {
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/db/src/lib.rs:70]: ✅

File: crates/client/db/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                    rocksdb_path: frontier_database_dir(db_config_dir, "db"),
+                    paritydb_path: frontier_database_dir(db_config_dir, "paritydb"),
+                    cache_size: 0,
+                },
+                _ => return Err("Supported db sources: `rocksdb` | `paritydb` | `auto`".to_string()),
+            },
+        })
+    }
+
+    pub fn new(config: &DatabaseSettings) -> Result<Self, String> {
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/db/src/lib.rs]: not public anymore

File: crates/client/db/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        let db = utils::open_database(config)?;
+
+        Ok(Self {
+            mapping: Arc::new(MappingDb { db: db.clone(), write_lock: Arc::new(Mutex::new(())), _marker: PhantomData }),
+            meta: Arc::new(MetaDb { db: db.clone(), _marker: PhantomData }),
+        })
+    }
+
+    pub fn mapping(&self) -> &Arc<MappingDb<Block>> {
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/db/src/lib.rs:99]: ✅

File: crates/client/db/src/mapping_db.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+// Substrate
+use scale_codec::{Decode, Encode};
+use sp_core::H256;
+use sp_database::Database;
+use sp_runtime::traits::Block as BlockT;
+
+use crate::DbHash;
+
+#[derive(Clone, Debug, Eq, PartialEq, Encode, Decode)]
+pub struct TransactionMetadata<Block: BlockT> {
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/db/src/mapping_db.rs]: removed so no need for comments

File: crates/client/db/src/mapping_db.rs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            _ => vec![commitment.block_hash],
+        };
+
+        transaction.set(
+            crate::columns::BLOCK_MAPPING,
+            &commitment.starknet_block_hash.encode(),
+            &substrate_hashes.encode(),
+        );
+
+        // TODO: add back when transaction support
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/db/src/mapping_db.rs:106]: ✅ 
no even sure we need it. Will confirm while unrolling the RPCs

File: crates/node/Cargo.toml [hunk 7]
```diff
@@ -63,8 +63,13 @@ substrate-frame-rpc-system = { workspace = true }
 frame-benchmarking = { workspace = true }
 frame-benchmarking-cli = { workspace = true }
 
-# Local Dependencies
+# Starknet
 madara-runtime = { path = "../runtime" }
+mc-db = { path = "../client/db" }
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/node/Cargo.toml:68]: ✅

**Comment by tdelabro** [comment]
I think it's ready for another review.
I added a lot of doc comments.
I agree with the lack of tests. The thing is that there are no tests for those things in Frontier either. Given that I mostly copy-pasted their code, I am confident it will work. Moreover, it does work when I run the chain. Adding tests on some of those crates could be an interesting first issue for somebody who wants to discover the codebase and the RPC flow.

I would be more than happy to do a live stream where I explain the RPC flow and the reason for those multiple crates to the project contributors. I will ask on the telegram channel if some would be interested and at what time.

File: crates/client/db/src/mapping_db.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (abdelhamidbakhta) [reply crates/client/db/src/mapping_db.rs]: let me know about it so that I can close this conversation

**Review by abdelhamidbakhta — APPROVED** [review]
lgtm

**Pull request merged**